### PR TITLE
Logtask

### PR DESCRIPTION
**Pull request by newEZ**
Rewrite the logging part of httpproxy and proxy. Remain work of testing the log in different cases thoroughly and change the hardcoded reply time of httpproxy. The new structure is:

IP address, name, type, class, qid, flags(sep by /), X/Y/Z, edns(sep by /)

for the query answer we also have those additional fields:

RCODE, time of reply

### CODE REVIEW

File: dohproxy/utils.py [hunk 1]
```diff
@@ -21,16 +21,48 @@
...
-        dns.flags.to_text(msg.flags),
+        flags,
+    )
+
+
+def dnsans2log(msg: dns.message.Message) -> str:
+    """ Helper function to return a readable excerpt from a dns answer object.
+    """
+    question = '<empty>'
+    if len(msg.question):
```
Comment (chantra) [inline dohproxy/utils.py]: same, also, if you use the same logic more than once, you should have a helper function.

File: dohproxy/utils.py [hunk 2]
```diff
@@ -21,16 +21,48 @@
...
+        qclass = questions[1]
+        qtype = questions[2]
+        question = ' '.join([qname, qtype, qclass])
+    flags = '/'.join(dns.flags.to_text(msg.flags).split(' '))
+
+    return '{} {} {} {}/{}/{} {}/{}/{} {}'.format(
         question,
+        msg.id,
+        flags,
+        len(msg.answer[0].items) if len(msg.answer) else 0,
```
Comment (chantra) [inline dohproxy/utils.py]: this is not correct, I believe this should be:
```
sum(len(x) for x in msg.answer)
```

The same applies for all sections, so feel free to use a helper function to sum items in a section.

File: dohproxy/proxy.py [hunk 3]
```diff
@@ -136,12 +140,14 @@ def stream_complete(self, stream_id: int):
             self.return_400(stream_id, body=e.body())
             return
 
+        clientip = self.transport.get_extra_info('peername')[0]
         self.logger.info(
-            '[HTTPS] Received: {} Peer {}'.format(
-                utils.dnsmsg2log(dnsq),
-                self.transport.get_extra_info('peername'),
+            '{} {}'.format(
```
Comment (chantra) [inline dohproxy/proxy.py]: We still need to keep the `[HTTPS]` or `[DNS]` and wether it is received or sent at the moment. This helps in understanding where/when the logs are being written. This applies for each log entries you have edited.

File: dohproxy/proxy.py [hunk 4]
```diff
@@ -136,12 +140,14 @@ def stream_complete(self, stream_id: int):
...
+                clientip,
+                utils.dnsquery2log(dnsq)
             )
         )
+        self.time_stamp = int(round(time.time() * 1000))
```
Comment (chantra) [inline dohproxy/proxy.py]: why not using the actual time?

File: dohproxy/proxy.py [hunk 5]
```diff
@@ -155,10 +161,13 @@ def on_answer(self, stream_id, dnsr=None, dnsq=None):
             ttl = min(r.ttl for r in dnsr.answer)
             headers['cache-control'] = 'max-age={}'.format(ttl)
 
+        clientip = self.transport.get_extra_info('peername')[0]
+        interval = int(round(time.time() * 1000)) - self.time_stamp
```
Comment (chantra) [inline dohproxy/proxy.py]: this would become:
```
interval = int((time.time() - self.time_stamp) * 1000)
```

File: dohproxy/server_protocol.py [hunk 6]
```diff
@@ -29,23 +29,37 @@ class DOHDNSException(DOHException):
...
         self.transport = transport
         self.dnsq.id = dns.entropy.random_16()
-        self.logger.info('[DNS] Send: {}'.format(utils.dnsmsg2log(self.dnsq)))
+        self.logger.info(
+            '{} {}'.format(
+                self.clientip,
+                utils.dnsquery2log(self.dnsq)
+            )
+        )
+        self.time_stamp = int(round(time.time() * 1000))
```
Comment (chantra) [inline dohproxy/server_protocol.py]: same comment. when it comes to timing.

**Comment by newEZ** [comment]
Change code according to the comments and change test to make sure it is up to date with the new code(because some functions change)

**Comment by chantra** [comment]
Thanks @newEZ , this looks good.

**Pull request merged**